reco_utils/recommender/rlrmc/RLRMCalgorithm.py

```diff
@@ -0,0 +1,224 @@
+from pymanopt.solvers.linesearch import LineSearchBackTracking
+from pymanopt.manifolds import Stiefel, PositiveDefinite, Product
+from math import sqrt
+from scipy.sparse import csr_matrix
+from scipy.sparse.linalg import svds
+import time 
+from numba import jit, njit, prange
+
+class RLRMCalgorithm(object):
+    """
+    classdocs
+    """
+
+    def __init__(
+        self,
+        rank,
+        C,
+        model_param,
+        initialize_flag = 'random',
+        max_time = 1000,
+        maxiter = 100,
+        seed = 42,
+    ):
+        """
+        Constructor
+        """
+        self.model_param = model_param
+        self.train_mean = model_param.get('train_mean')
+        self.initialize_flag = initialize_flag
+        self.rank = rank
+        self.C = C
+        self.max_time = max_time
+        self.maxiter = maxiter
+
+    def _init_train(self,entries_train_csr):
+        print("Hyper-parameters of the algorithm")
+        print("Rank: %i, Regularization parameter: %e" % (self.rank, self.C))
+        # Initialization # starting point on the manifold
+        initialize_dict = {'random': 0, 'svd': 1}
+        if initialize_dict.get(self.initialize_flag)==0:#rndom
+            W0=None
+        elif initialize_dict.get(self.initialize_flag)==1:#svd
+            U0, B0, V0 = svds(entries_train_csr, k=self.rank)
+            W0 = [U0, V0.T, np.diag(B0)]
+        else: # default option when given incorrect option
+            print("Initialization flag not recognized. Setting it to random (default).")
+            W0=None
+        return W0
+
+    def fit(self,RLRMCdata,verbosity = 0,compute_iter_rmse = False):
+        # initialize the model
+        W0 = self._init_train(RLRMCdata.train)
+        self.user2id = RLRMCdata.user2id
+        self.item2id = RLRMCdata.item2id
+        self.id2user = RLRMCdata.id2user
+        self.id2item = RLRMCdata.id2item
+
+        # residual variable
+        residual_global = np.zeros(RLRMCdata.train.data.shape,dtype=np.float64)
+
+        ###################Riemannian first-order algorithm######################
+
+        solver = ConjugateGradientMS(maxtime=self.max_time, 
+            maxiter=self.maxiter, linesearch=LineSearchBackTracking())#, logverbosity=2)
+        # construction of manifold
+        manifold = Product([
+            Stiefel(self.model_param.get('num_row'), self.rank), 
+            Stiefel(self.model_param.get('num_col'), self.rank), 
+            PositiveDefinite(self.rank)
+            ])
+        problem = Problem(
+            manifold=manifold, 
+            cost=lambda x: self.cost(x,RLRMCdata.train.data,RLRMCdata.train.indices,RLRMCdata.train.indptr,residual_global), 
+            egrad=lambda z: self.egrad(z,RLRMCdata.train.indices,RLRMCdata.train.indptr,residual_global), 
+            verbosity=verbosity)
+        
+        residual_test_global = (np.zeros(RLRMCdata.test.data.shape,dtype=np.float64))
+        Wopt, self.stats = solver.solve(problem, x=W0, 
```

Comment:
This function does evaluate on test data (at my_stats) if the test data is set. Would be better to split functions into fit / evaluation / (and fit_and_evaluate if you want) to have single responsibility.

We may want to track the evaluation performance while training, like using a training hook. If this is the case,  would be great to use the term **validation** instead of **test** here and specify how to use that in the docstring.

Reply:
Done. Added an additional fit_evaluate function and changed the "test" terms to "validation" terms.